core/src/raw/oio/read/buffer_reader.rs

```diff
@@ -0,0 +1,469 @@
+where
+    R: oio::Read,
+{
+    fn poll_fill_buf(&mut self, cx: &mut Context<'_>) -> Poll<Result<&[u8]>> {
+        // If we've reached the end of our internal buffer then we need to fetch
+        // some more data from the underlying reader.
+        // Branch using `>=` instead of the more correct `==`
+        // to tell the compiler that the pos..cap slice is always valid.
+        if self.pos >= self.filled {
+            debug_assert!(self.pos == self.filled);
+
+            let cap = self.capacity();
+            self.buf.clear();
+            let dst = self.buf.spare_capacity_mut();
+            let mut buf = ReadBuf::uninit(dst);
+            unsafe { buf.assume_init(cap) };
+
+            let n = ready!(self.r.poll_read(cx, buf.initialized_mut()))?;
+            unsafe { self.buf.set_len(n) }
+
+            self.pos = 0;
+            self.filled = n;
+        }
+
+        Poll::Ready(Ok(&self.buf[self.pos..self.filled]))
+    }
+
+    fn consume(&mut self, amt: usize) {
+        self.pos = min(self.pos + amt, self.filled);
+        self.cur += amt as u64;
+    }
+
+    fn seek_relative(&mut self, offset: i64) -> Option<u64> {
+        let pos = self.pos as u64;
+        if offset < 0 {
+            if let Some(new_pos) = pos.checked_sub((-offset) as u64) {
+                self.cur -= (-offset) as u64;
+                self.pos = new_pos as usize;
+                return Some(self.cur);
+            }
+        } else if let Some(new_pos) = pos.checked_add(offset as u64) {
+            if new_pos <= self.filled as u64 {
+                self.cur += offset as u64;
+                self.pos = new_pos as usize;
+                return Some(self.cur);
+            }
+        }
+
+        None
+    }
+
+    fn poll_inner_seek(&mut self, cx: &mut Context<'_>, pos: SeekFrom) -> Poll<Result<u64>> {
+        let cur = ready!(self.r.poll_seek(cx, pos))?;
+        self.discard_buffer();
+        self.cur = cur;
+
+        Poll::Ready(Ok(cur))
+    }
+}
+
+impl<R> oio::Read for BufferReader<R>
+where
+    R: oio::Read,
+{
+    fn poll_read(&mut self, cx: &mut Context<'_>, mut dst: &mut [u8]) -> Poll<Result<usize>> {
+        // Sanity check for normal cases.
+        if dst.is_empty() {
+            return Poll::Ready(Ok(0));
+        }
+
+        // If we don't have any buffered data and we're doing a massive read
+        // (larger than our internal buffer), bypass our internal buffer
+        // entirely.
+        if self.pos == self.filled && dst.len() >= self.capacity() {
+            let res = ready!(self.r.poll_read(cx, dst));
+            self.discard_buffer();
+            return Poll::Ready(res);
+        }
+
+        let rem = ready!(self.poll_fill_buf(cx))?;
+
+        let amt = min(rem.len(), dst.len());
+        dst.put(&rem[..amt]);
+        self.consume(amt);
+        Poll::Ready(Ok(amt))
+    }
+
+    fn poll_seek(&mut self, cx: &mut Context<'_>, pos: SeekFrom) -> Poll<Result<u64>> {
+        let offset = match pos {
```

Comment:
Maybe we don't need to check offset? It should be safe to forward `poll_inner_seek` directly and let underlying reader to return the `InvalidInput` error. BufferReader doesn't change this behavior.